File: tests/src/com/g3ida/withflyingcolours/tests/RotationSystemTest.java
package com.g3ida.withflyingcolours.tests;

import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.g3ida.withflyingcolours.Utils;
import com.g3ida.withflyingcolours.core.player.movement.PlayerRotationComponent;
import com.g3ida.withflyingcolours.core.player.movement.PlayerRotationSystem;
import com.g3ida.withflyingcolours.utils.RotationDirection;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;

import games.rednblack.editor.renderer.components.physics.PhysicsBodyComponent;
import games.rednblack.editor.renderer.utils.ComponentRetriever;

@RunWith(GdxTestRunner.class)
public class RotationSystemTest extends BaseSystemTests {
    @Test
    public void testPlayerRotationSystem() throws AssertionError {

        //prepare engine and entity with physics
        Engine engine = createPhysicsWorldWithDynamicEntity();
        Entity entity = engine.getEntities().first();

        engine.addSystem(new PlayerRotationSystem());
        ComponentRetriever.addMapper(PlayerRotationComponent.class);
        entity.add(new PlayerRotationComponent());

        PhysicsBodyComponent PhysicsBodyComponent = ComponentRetriever.get(entity, PhysicsBodyComponent.class);
        PlayerRotationComponent playerRotationComponent = ComponentRetriever.get(entity, PlayerRotationComponent.class);

        // check that the entity have a valid body and its initial rotation set to 0.
        cycleEngineFor(engine, 1);
        Assert.assertNotNull(PhysicsBodyComponent.body);
        PhysicsBodyComponent.body.setFixedRotation(true);
        Assert.assertEquals(0f, PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the right by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.clockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(-Utils.PI2, PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the right by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.clockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(-Utils.PI, PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the right by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.clockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(-(Utils.PI + Utils.PI2), PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the right by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.clockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(0f, Utils.modPI2(PhysicsBodyComponent.body.getAngle()), 0.01f);

        // rotate the entity to the left by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.antiClockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(-(Utils.PI + Utils.PI2), PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the left by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.antiClockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(-Utils.PI, PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the left by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.antiClockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(-Utils.PI2, PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the left by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.antiClockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(0, PhysicsBodyComponent.body.getAngle(), 0.01f);

        // rotate the entity to the left by 90deg and assert.
        playerRotationComponent.setRotationDirection(RotationDirection.antiClockwise);
        cycleEngineFor(engine, 1);
        Assert.assertEquals(Utils.PI2, PhysicsBodyComponent.body.getAngle(), 0.01f);

        //TODO: add test cases with multiple instantaneous rotate request and with variable speed etc...
    }
}

File: tests/src/com/g3ida/withflyingcolours/tests/RotationSystemTest.kt
package com.g3ida.withflyingcolours.tests

import com.g3ida.withflyingcolours.core.common.GameSettings
import com.g3ida.withflyingcolours.core.ecs.components.EventListenerComponent
import com.g3ida.withflyingcolours.core.ecs.systems.EventListenerSystem
import com.g3ida.withflyingcolours.core.events.EventType
import com.g3ida.withflyingcolours.core.events.GameEvent
import com.g3ida.withflyingcolours.utils.extensions.PI
import com.g3ida.withflyingcolours.utils.extensions.PI2
import com.g3ida.withflyingcolours.utils.extensions.modPI2
import com.g3ida.withflyingcolours.core.actions.PlayerRotationAction
import com.g3ida.withflyingcolours.core.actions.toActionListener
import games.rednblack.editor.renderer.components.physics.PhysicsBodyComponent
import org.junit.runner.RunWith
import ktx.collections.gdxMapOf
import org.junit.Assert
import org.junit.Test
import java.lang.AssertionError

@RunWith(GdxTestRunner::class)
class RotationSystemTest : BaseSystemTests() {
    @Test
    @Throws(AssertionError::class)
    fun testPlayerRotationSystem() {
        //prepare engine and entity with physics
        val engineWithEntity = createPhysicsWorldWithDynamicEntity(EventListenerSystem())
        val engine = engineWithEntity.world
        val entityId = engineWithEntity.entityId
        val playerRotationComponent = engine.edit(entityId).create(EventListenerComponent::class.java)
        val physicsBodyComponent = engine.getMapper(PhysicsBodyComponent::class.java)[entityId]

        // check that the entity have a valid body and its initial rotation set to 0.
        cycleEngineFor(engine, 1f)
        Assert.assertNotNull(physicsBodyComponent.body)
        physicsBodyComponent.body.isFixedRotation = true
        Assert.assertEquals(0f, physicsBodyComponent.body.angle, 0.01f)

        playerRotationComponent.addActionListener(PlayerRotationAction(physicsBodyComponent).toActionListener(EventType.RotateCommand))

        // rotate the entity to the right by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "-1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(-Float.PI2, physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the right by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "-1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(-Float.PI, physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the right by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "-1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(-(Float.PI + Float.PI2), physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the right by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "-1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(0f, physicsBodyComponent.body.angle.modPI2(), 0.01f)

        // rotate the entity to the left by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(-(Float.PI + Float.PI2), physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the left by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(-Float.PI, physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the left by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(-Float.PI2, physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the left by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(0f, physicsBodyComponent.body.angle, 0.01f)

        // rotate the entity to the left by 90deg and assert.
        GameSettings.eventHandler.dispatchEvent(GameEvent(EventType.RotateCommand, gdxMapOf("direction" to "1")))
        cycleEngineFor(engine, 1f)
        Assert.assertEquals(Float.PI2, physicsBodyComponent.body.angle, 0.01f)

        //TODO: add test cases with multiple instantaneous rotate request and with variable speed etc...
    }
}